crates/rust-analyzer/src/config.rs

```diff
@@ -2284,11 +2320,20 @@ mod single_or_array {
     }
 }
 
-#[derive(Serialize, Deserialize, Debug, Clone)]
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
 #[serde(untagged)]
 enum ManifestOrProjectJson {
     Manifest(Utf8PathBuf),
-    ProjectJson(ProjectJsonData),
+    ProjectJson { data: ProjectJsonData, buildfile: Option<Utf8PathBuf> },
```

Comment:
oh, i just realized that this change alone would break (de)serialization for callers to want to pass the `rust-project.json` inline.